src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  for (let i = 0; i < actions.length; i++) {
+    const action = actions[i];
+
+    switch (action.type) {
+      case 'clear':
+        stateCopy = {};
+        break;
+      case 'addProperties':
+        stateCopy = { ...stateCopy, ...action.extraData };
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+      default:
+        break;
+    }
+
+    stateHistory.push({ ...stateCopy });
```

Comment:
Consider pushing the copy of the state object to the `stateHistory` array at the end of the loop, but outside of the `switch` block. This way, the action of adding the state to the history is not repeated in every case and adheres to the DRY (Don't Repeat Yourself) principle. It's already correctly placed outside of the `switch`, but make sure to maintain this practice in future implementations.